### DISCUSSION TITLE
convert RunTime::Buffer to Halide::Buffer

### DISCUSSION THREAD

**Discussion by hailangWu-Aval** [halide/Halide#5266] [Q&A]
Is that any different between Halide::RunTime::Buffer and Halide::Buffer ?  

how can i convert  RunTime::Buffer to Halide::Buffer  or Halide::Buffer to RunTime::Buffer?

**Comment by abadams** [comment]
A Halide::Buffer is a named shared pointer to a Halide::Runtime::Buffer

To convert from a Halide::Runtime::Buffer to a Halide::Buffer you do something like:

```
Halide::Buffer<float> buf(std::move(my_halide_runtime_buffer), "some name")
```

To convert back, Halide::Buffer has a get() method that returns a pointer to the underlying Runtime::Buffer.

**Comment by hailangWu-Aval** [comment]
that's cool! thanks a lot.

there is an another little question about the efficiency of converting interleaved to planar. i have compared halide func method and traditional method with pointer assginment, it turns out the func way cost twice than the point assignment method. So is that has a more effective way to make it happend?

##1. halide func
 template <class T>
Halide::Buffer<T> interleaved2planar(Halide::Buffer<T> in) {
    Halide::ImageParam src(in.type(), 3);
    Halide::Func plane("plane");
    Halide::Var x, y, c;

    plane(x, y, c) = src(x, y, c);

    src.dim(0).set_stride(3).dim(2).set_stride(1).set_bounds(0, 3);
    plane.output_buffer().dim(0).set_stride(1).dim(2).set_extent(3);

    plane.reorder(c, x, y).unroll(c);
    // plane.vectorize(x, 64);

    src.set(in);
    plane.compile_jit(); // it works with or without compile_jit

    auto out_buf = Halide::Buffer<T>(in.width(), in.height(), 3);
    plane.realize(out_buf);
    return out_buf;
}

##2. pointer assignment 
void interleaved2planar(const cv::Mat& input, float* data)
{
    int w = input.cols;
    int h = input.rows;
    int c = input.channels();

    const float *ptr_input;
    int index = 0;
    auto ptr_data_1 = data;
    auto ptr_data_2 = data + h * w;
    auto ptr_data_3 = data + 2 * h * w;
    for (int row = 0; row < h; row++)
    {
        ptr_input = input.ptr<float>(row);
        
        for (int col = 0; col < w*c; col=col+3)
        {
            *(ptr_data_1) = *(ptr_input + col + 0) ;
            *(ptr_data_2) = *(ptr_input + col + 1) ;
            *(ptr_data_3) = *(ptr_input + col + 2) ;
            ++ptr_data_1;
            ++ptr_data_2;
            ++ptr_data_3;
        }
    }
}
